### PR TITLE
Update some outdated descriptions of coverage instrumentation

### PR DESCRIPTION
**Pull request by Zalathar**
Coverage internals have changed quite a bit over recent months/weeks, so I decided to at least make a pass over the existing docs and update anything that is clearly outdated.

### CODE REVIEW

**Comment by JohnTitor** [comment]
I'm not familiar with that, could you point to a relevant PR (or ping a better reviewer)?

**Comment by Zalathar** [comment]
This is mainly based on my own recent work in https://github.com/rust-lang/rust/pull/116046.

It used to be that mappings between coverage counters and source code regions were stored directly in special coverage statements that the coverage instrumentor would inject into MIR.

After that change, those mappings are now stored in a `FunctionCoverageInfo` struct that is attached to the function's `mir::Body`.

---

The part about injecting counters into MIR has been obsoleted by follow-up work in https://github.com/rust-lang/rust/pull/116917.

What used to be a fairly intricate region of code for injecting various things into MIR has been replaced by two fairly simple loops.

File: src/llvm-coverage-instrumentation.md [hunk 1]
```diff
@@ -327,9 +326,10 @@ Instrumentor::new(&self.name(), tcx, mir_body).inject_counters();
 The `CoverageGraph` is a coverage-specific simplification of the MIR control
 flow graph (CFG). Its nodes are [`BasicCoverageBlock`s][bcb], which
 encompass one or more sequentially-executed MIR `BasicBlock`s
-(with no internal branching), plus a `CoverageKind` counter (to
-be added, via coverage analysis), and an optional set of additional counters
-to count incoming edges (if there are more than one).
```
Comment (Zalathar) [inline src/llvm-coverage-instrumentation.md:332]: The extra fields mentioned here were moved out into side-tables in `CoverageCounters` in https://github.com/rust-lang/rust/pull/114354.

File: src/llvm-coverage-instrumentation.md [hunk 2]
```diff
@@ -338,16 +338,15 @@ with the following steps:
...
+   `BcbCounter::Expression`s for each `CoverageSpan`, plus additional
+   _intermediate expressions_[^intermediate-expressions] that are not associated
+   with any `CodeRegion`, but
    are required to compute a final `Expression` value for a `CodeRegion`.
 3. Inject the new counters into the MIR, as new `StatementKind::Coverage`
-   statements. This is done by three distinct functions:
-   - `inject_coverage_span_counters()`
-   - `inject_indirect_counters()`
-   - `inject_intermediate_expression()`, called for each intermediate expression
-     returned from `make_bcb_counters()`
```
Comment (Zalathar) [inline src/llvm-coverage-instrumentation.md:350]: The three `inject` function mentioned here were removed by https://github.com/rust-lang/rust/pull/116917, and replaced with a much simpler pair of loops in `create_mappings_and_inject_coverage_statements`.

**Review by Swatinem — APPROVED** [review]

**Review by davidtwco — APPROVED** [review]

**Pull request merged**